ExamenProeveUnity/Assets/Scripts/Runtime/Shelf.cs

```diff
@@ -1,29 +1,164 @@
+using System;
 using System.Collections.Generic;
+using System.Security.Cryptography;
 using Runtime.Interfaces;
+using TMPro;
+using Toolbox.MethodExtensions;
 using UnityEngine;
+using UnityEngine.Events;
+using UnityEngine.Serialization;
+using Utilities.MethodExtensions;
 
-public class Shelf : MonoBehaviour, IGridable
+namespace Runtime
 {
-    [Tooltip("The index of the grid node where the player should stand to interact with the shelf")]
-    private readonly List<ProductScriptableObject> _inventory = new List<ProductScriptableObject>() { Capacity = 5 };
+    public class Shelf : MonoBehaviour, IGridable
+    {
+        [SerializeField] private int maxInventorySize = 5;
+        [SerializeField] private ProductScriptableObject _item;
+        [SerializeField] private List<GameObject> _displayedItems = new List<GameObject>();
+        [SerializeField] private TMP_Text itemCountText;
+        [SerializeField] private Canvas canvas;
+    
+        [Tooltip("The index of the grid node where the player should stand to interact with the shelf")]
+        [field: SerializeField] 
+        public int gridIndex { get; set; }
+    
+        public int InteractionGridIndex => gridIndex;
+        public ProductScriptableObject Item => _item;
+        public int ItemCount => _itemCount;
     
-    [field: SerializeField] public int gridIndex { get; set; }
-    public int InteractionGridIndex => gridIndex;
+        private int _itemCount;
+    
+        public UnityEvent onItemAdded = new();
+        public UnityEvent onItemRemoved = new();
+        public UnityEvent onInventorySizeChange = new();
 
-    public void Restock(ProductScriptableObject item)
-    {
-        for (int i = _inventory.Count; i < _inventory.Capacity; i++)
+
+        private void Awake()
         {
-            _inventory.Add(item);
+            _displayedItems.ForEach(item => item.SetActive(false));
+            for (int i = 0; i < _itemCount; i++)
+            {
+                if(!_displayedItems.ContainsSlot(i)) continue;
+                _displayedItems[i].SetActive(true);
+            }
+
+            if (canvas != null)
+            {
+                canvas.enabled = false;
+            }
+            
+            if(itemCountText != null)
+            {
+                itemCountText.text = _itemCount.ToString();
+            }
         }
-    }
 
-    public ProductScriptableObject GrabItem()
-    {
-        if (_inventory.Count < 1) return null;
-        ProductScriptableObject productToReturn = _inventory[0];
-        _inventory.RemoveAt(0);
-        return productToReturn;
-    }
+        /// <summary>
+        /// Fulls the shelf with the given item
+        /// </summary>
+        public void Full()
```

Comment:
naam zou 'Fill' moeten zijn
Full zou eerder een true/false returnen als de shelf full is of niet